File: code/0993 Cousins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isCousins(self, root, x, y):
        """
        :type root: TreeNode
        :type x: int
        :type y: int
        :rtype: bool
        """
        g = {}
        def f(root, i=0, parentVal=None):
            if root == None:
                return
            g[root.val] = (i, parentVal)
            f(root.left, i+1, parentVal=root.val)
            f(root.right, i+1, parentVal=root.val)
        f(root)
        return g[x][0] == g[y][0] and g[x][1] != g[y][1]
        

File: code/0993 Cousins in Binary Tree.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void FindDepthandParent(struct TreeNode* root,int x,int depth,int parent,int *depthx,int *parentx){
    if (!root) return;
    if (root->val==x){
        *depthx=depth;
        *parentx=parent;
        return;
    }
    FindDepthandParent(root->left,x,depth+1,root->val,depthx,parentx);
    FindDepthandParent(root->right,x,depth+1,root->val,depthx,parentx);
}
bool isCousins(struct TreeNode* root, int x, int y){
    if(!root) return false;
    int depthx=0;int depthy=0;
    int parentx=root->val;int parenty=root->val;
    FindDepthandParent(root,x,0,0,&depthx,&parentx);
    FindDepthandParent(root,y,0,0,&depthy,&parenty);
    return depthx==depthy&&parentx!=parenty;
}
    
